### PR TITLE
[KYUUBI #4881] JDBCEngine performs initialization sql

### PR DESCRIPTION
**Pull request by rhh777**
### _Why are the changes needed?_

close https://github.com/apache/kyuubi/issues/4881
### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/JdbcDialect.scala [hunk 1]
```diff
@@ -65,6 +65,8 @@ abstract class JdbcDialect extends SupportServiceLoader with Logging {
 
   def getSchemaHelper(): SchemaHelper
 
+  def initializationSQLs(): Seq[String]
```
Comment (pan3793) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/JdbcDialect.scala:68]: why not give it a default implement `SELECT 1`? then we don't need to overwrite it in each derived classes.

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/util/KyuubiJdbcUtil.scala [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import scala.util.{Failure, Success, Try}
+
+import org.apache.kyuubi.{KyuubiSQLException, Logging}
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.engine.jdbc.connection.ConnectionProvider
+import org.apache.kyuubi.engine.jdbc.dialect.{JdbcDialect, JdbcDialects}
+
+object KyuubiJdbcUtil extends Logging {
+
+    def initializeJdbcSession(kyuubiConf: KyuubiConf): Unit = {
```
Comment (pan3793) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/util/KyuubiJdbcUtil.scala:31]: - if a method is used in only one place, creating a `Util` object is not a good practice.

- Kyuubi uses `AbcUtils` style, not `AbcUtil`.

- `org.apache.kyuubi.util.JdbcUtils#withCloseable` can simplify code

Comment (rhh777) [reply externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/util/KyuubiJdbcUtil.scala:31]: > * if a method is used in only one place, creating a `Util` object is not a good practice.
> * Kyuubi uses `AbcUtils` style, not `AbcUtil`.
> * `org.apache.kyuubi.util.JdbcUtils#withCloseable` can simplify code
* emm, I am considering whether to add a method to execute Session Init SQL in Util
* I refer to the **KyuubiSparkUtil** format before, I will update later
* I'll look at the **JdbcUtils** code in more detail later

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/JdbcDialect.scala (hunk 1 above)
Comment (rhh777) [reply externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/JdbcDialect.scala:68]: > why not give it a default implementation `SELECT 1`? then we don't need to overwrite it in each derived classes.

It's a good idea to default to SELECT 1, updated later

**Comment by pan3793** [comment]
BTW, we provide `dev/reformat` to quickly fix the style.

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/JdbcDialect.scala [hunk 3]
```diff
@@ -65,6 +65,10 @@ abstract class JdbcDialect extends SupportServiceLoader with Logging {
 
   def getSchemaHelper(): SchemaHelper
 
+  def initializationSQLs(): Seq[String] = {
```
Comment (bowenliang123) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/JdbcDialect.scala:68]: The initialization SQLs should be configurable.

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/util/KyuubiJdbcUtils.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import org.apache.kyuubi.engine.jdbc.dialect.{JdbcDialect, JdbcDialects}
+import org.apache.kyuubi.util.JdbcUtils
+
+object KyuubiJdbcUtils extends Logging {
+
+  def initializeJdbcSession(kyuubiConf: KyuubiConf): Unit = {
+    try {
+      val dialect: JdbcDialect = JdbcDialects.get(kyuubiConf)
+      JdbcUtils.withCloseable(ConnectionProvider.create(kyuubiConf)) { connection =>
+        JdbcUtils.withCloseable(dialect.createStatement(connection, 100)) { statement =>
```
Comment (bowenliang123) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/util/KyuubiJdbcUtils.scala:32]: The timeout for executing init SQL should be configurable.

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PhoenixDialect.scala [hunk 5]
```diff
@@ -150,4 +150,5 @@ class PhoenixDialect extends JdbcDialect {
   override def name(): String = {
     "phoenix"
   }
+
```
Comment (bowenliang123) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PhoenixDialect.scala:153]: Remove the added blank line.

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/DorisDialect.scala [hunk 6]
```diff
@@ -162,4 +162,5 @@ class DorisDialect extends JdbcDialect {
   override def name(): String = {
     "doris"
   }
+
```
Comment (bowenliang123) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/DorisDialect.scala:165]: ditto.

Comment (rhh777) [reply externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/DorisDialect.scala]: > ditto.

Updated.

**Review by pan3793 — APPROVED** [review]
The implementation looks good to me, but I don't think the code should be put into `KyuubiJdbcUtils`, the code is not reusable totally, and it is compose of specific logic, not generic enough to put into a `*Utils`.

I would suggest move the `initializeJdbcSession` into `JdbcSQLEngine` directly. But if you consist, I'm fine to merge it as-is.

**Comment by rhh777** [comment]
> The implementation looks good to me, but I don't think the code should be put into `KyuubiJdbcUtils`, the code is not reusable totally, and it is compose of specific logic, not generic enough to put into a `*Utils`.
> 
> I would suggest move the `initializeJdbcSession` into `JdbcSQLEngine` directly. But if you consist, I'm fine to merge it as-is.

Do you mean that `KyuubiJdbcUtils` methods are all move to `JdbcSQLEngine`? `JdbcSessionImpl open()` calls the `initializeJdbcSession` method of `JdbcSQLEngine`?

**Comment by pan3793** [comment]
Yes, all things in `KyuubiJdbcUtils`. And the second method can be inlined, I don't see the benefit of extracting it as a method.

**Comment by rhh777** [comment]
> Yes, all things in `KyuubiJdbcUtils`. And the second method can be inlined, I don't see the benefit of extracting it as a method.

After `JdbcSQLEngine` is initialized, the connection can be closed. However, after `JdbcSessionImpl open()` is initialized, the connection cannot be closed and will be used by the sql of subsequent users. So the `initializeJdbcSession` method is split.

**Review by cxzl25 — CHANGES_REQUESTED** [review]
Overall LGTM.
Please use the following command to update `settings.md`
```bash
KYUUBI_UPDATE=1 build/mvn clean test -pl kyuubi-server -am -Pflink-provided,spark-provided,hive-provided -Dtest=none -DwildcardSuites=org.apache.kyuubi.config.AllKyuubiConfiguration
```

**Comment by rhh777** [comment]
> Overall LGTM. Please use the following command to update `settings.md`
> 
> ```shell
> KYUUBI_UPDATE=1 build/mvn clean test -pl kyuubi-server -am -Pflink-provided,spark-provided,hive-provided -Dtest=none -DwildcardSuites=org.apache.kyuubi.config.AllKyuubiConfiguration
> ```

Updated

**Review by cxzl25 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master

**Pull request closed without merging**